include/rear-panel/core/lights/animation_handler.hpp

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <algorithm>
+#include <optional>
+
+#include "rear-panel/core/lights/animation_math.hpp"
+#include "rear-panel/core/lights/animation_queue.hpp"
+
+namespace lights {
+
+/**
+ * @brief Acts as an interface for an AnimationQueue
+ *
+ */
+template <size_t Size>
+class AnimationHandler {
+  public:
+    AnimationHandler()
+        : _step_starting_color{.r = 0, .g = 0, .b = 0, .w = 0},
+          _most_recent_color{.r = 0, .g = 0, .b = 0, .w = 0},
+          _current_step{std::nullopt},
+          _timer_ms(0),
+          _queue() {}
+
+    /**
+     * @brief Advance the animation and get the next color setting.
+     *
+     * @param time_increment_ms The amount of time that has passed, in
+     * milliseconds.
+     * @return Color
+     */
+    auto animate(uint32_t time_increment_ms) -> Color {
+        Color ret = _step_starting_color;
+        if (!_current_step.has_value()) {
+            // We were not in an active animation last update, but
+            // one may have started since then. Try to get a new step.
+            _timer_ms = 0;
+            _current_step = _queue.get_next_active_step();
+        }
+
+        if (_current_step.has_value()) {
+            // We are in a step, calculate the next increment
+            _timer_ms += time_increment_ms;
```

Comment:
The issue with that approach is that it ends up repeating states. Example, if you have two steps that are 3ms long and we tick at 1ms:

- First step: [0, 1, 2, 3]
- Second step [0, 1, 2, 3]

But the state of the first step at 3 is the same as the second step at 0. 

Since we're now [only grabbing](https://github.com/Opentrons/ot3-firmware/pull/617#discussion_r1146214499) from the queue at the start of the `animate` function, you end the step when the counter equals the total transition time but then the next time through you have to start at time = 0.

Agree that, either way, it should look the same to any human.

Reply:
Ahhhhh actually, after the pseudo-visitor refactor you suggested, we can do the following:
* When grabbing a new step from the queue, set the timer to the increment amount instead of 0
* Increment the timer _after_ calculating the interpolation

And we get the correct behavior without skipping anything. Calling `animate` gives the values for the _current_ timer and then advances it.